### PR TITLE
Upgrade to terraform 0.12

### LINKED ISSUE #376: add Terraform 0.12 support
# I have issues

## I'm submitting a...

* [ ] bug report
* [x] feature request
* [ ] support request
* [ ] kudos, thank you, warm fuzzy

## What is the current behavior?

Currently you cannot use this module with Terraform >= 0.12.0

There are several errors relating to:
  "Error: Unsupported block type"
  "Error: Missing resource instance key"

## Environment details

* Affected module version: 
* OS: NA
* Terraform version: v0.12.0-rc1

### PR DESCRIPTION
**Pull request by nauxliu**
## Description

Add Terraform 0.12 support, resolve #376

This is a fork of [alex-goncharov's fork](https://github.com/alex-goncharov/terraform-aws-eks) and resolved conflicts

### Remove `xx_count` variables from your code
 * map_accounts_count
 * map_roles_count
 * map_users_count
 * worker_group_count
 * worker_group_launch_template_mixed_count
 * worker_group_launch_template_count
 * workers_additional_policies_count

### Replace strings with commas with lists

Affected attributes:
  * subnets
  * additional_security_group_ids
  * suspended_processes
  * target_group_arns
  * enabled_metrics
  * termination_policies

example:

```diff
worker_groups = [
  {
    name    = "example"
-   subnets = "subnet-123,subnet-456,subnet-789"
+   subnets = ["subnet-123", "subnet-456", "subnet-789"]
  }
]
```

### Update worker group tags

```diff
 worker_groups = [
   {
     name = "example"
+    tags = [
+      {
+        key = "foo"
+        value = "bar"
+        propagate_at_launch = true
+      },
+    ]
   },
 ]

- worker_group_tags = {
-   default = []
-   example = [
-     {
-       key = "foo"
-       value = "bar"
-       propagate_at_launch = true
-     },
-   ]
- }
```

### Update worker_groups_launch_template_mixed override instance types

```diff
worker_groups_launch_template_mixed = [
  {
    name                     = "spot-1"
+   override_instance_types  = ["m5.large", "c5.large", "t3.large", "r5.large"]
-   override_instance_type_1 = "m5.large"
-   override_instance_type_2 = "c5.large"
-   override_instance_type_3 = "t3.large"
-   override_instance_type_4 = "r5.large"
    spot_instance_pools      = 3
    asg_max_size             = 5
    asg_desired_size         = 5
    autoscaling_enabled      = true
    kubelet_extra_args       = "--node-labels=kubernetes.io/lifecycle=spot"
  }
]
```

### CODE REVIEW

File: examples/launch_templates/main.tf [hunk 1]
```diff
@@ -24,9 +24,8 @@ resource "random_string" "suffix" {
 }
 
 module "vpc" {
-  #source         = "terraform-aws-modules/vpc/aws"
-  #version        = "1.60.0"
-  source = "git::https:///github.com/alex-goncharov/terraform-aws-vpc.git?ref=terraform-0.12"
+  source         = "terraform-aws-modules/vpc/aws"
+  version        = "2.5.0"
```
Comment (mathuin) [inline examples/launch_templates/main.tf]: v2.6.0 was released this morning.

**Comment by mathuin** [comment]
v2.6.0 of terrraform-aws-vpc was released this morning, so you might want to update the example.

**Comment by mathuin** [comment]
Also, possibly an unfair ask, but #338 adds some very useful functionality to v0.11 that would be super helpful in v0.12.

**Comment by nauxliu** [comment]
> Also, possibly an unfair ask, but #338 adds some very useful functionality to v0.11 that would be super helpful in v0.12.

@mathuin  #338 already merged, what can I do?

File: CHANGELOG.md [hunk 2]
```diff
@@ -18,6 +18,8 @@ project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Changed
 
+- Finally, Terraform 0.12 support (by @alex-goncharov @nauxliu @timboven)
+- All the xx_count variables have been removed (by @nauxliu)
```
Comment (max-rocket-internet) [inline CHANGELOG.md]: 💙

File: data.tf [hunk 3]
```diff
@@ -1,4 +1,5 @@
-data "aws_region" "current" {}
+data "aws_region" "current" {
+}
```
Comment (max-rocket-internet) [inline data.tf]: Does `terraform fmt` in 0.12 make this change?

Comment (nauxliu) [reply data.tf]: `terraform  0.12upgrade` does

**Comment by max-rocket-internet** [comment]
Some travis config is broken I think as it says `build passed` but the log is full of errors

**Comment by nauxliu** [comment]
> Some travis config is broken I think as it says `build passed` but the log is full of errors

**Comment by nauxliu** [comment]
> Some travis config is broken I think as it says `build passed` but the log is full of errors

Fixed, Ready for review.

**Comment by max-rocket-internet** [comment]
Thanks @nauxliu 🎉

It re-ordered the tags on all my resources but that was all. I'm using one `worker_groups_launch_template` worker group and one `worker_groups_launch_template_mixed`.

Let's get some more people to test before merging though.

**Comment by max-rocket-internet** [comment]
Everyone, please test this PR so we can merge it 😃
 https://github.com/terraform-aws-modules/terraform-aws-eks/pull/394

**Comment by markcornick** [comment]
Tested, works - like @max-rocket-internet I saw some reordering of tags, but nothing destructive. I was previously using the alex-goncharov fork.

**Comment by mathuin** [comment]
Fair warning:  I'm a newbie when it comes to terraform.

I'm getting some errors here in some recently-inherited code with this fork:

```
Error: Unsupported argument

  on main.tf line 143, in module "eks":
 143:   worker_group_launch_template_count   = "1"

An argument named "worker_group_launch_template_count" is not expected here.


Error: Unsupported argument

  on main.tf line 144, in module "eks":
 144:   worker_group_count                   = "0"

An argument named "worker_group_count" is not expected here.
```

Those two lines are in the module declaration, and appear to be the dreaded `xx_count` variables which were stripped out.  While I accept that they're not desirable, some kind of documentation explaining how to fix code containing them would be desirable.  They also still appear in the spot instance docs, so those might need some kind of change as well.

**Comment by nauxliu** [comment]
@mathuin The upgrade guide has been added to the issue description.

**Comment by nauxliu** [comment]
@mathuin @max-rocket-internet @markcornick 
I've made two more breaking changes, Could you test again?

**Comment by sc250024** [comment]
Thanks @nauxliu for taking care of this!

File: data.tf (hunk 3 above)
Comment (eytanhanig) [reply data.tf]: `terraform fmt` for `0.12.2` appears to be satisfied with either syntax. `terraform -0.12upgrade` has some unexpected opinions, like ending each file with a double line break, which `terraform fmt` also appears neutral on.

**Comment by alex-goncharov** [comment]
Thans @nauxliu, works for me.

**Comment by cawilliamson** [comment]
@nauxliu This is a little off-topic but is there any simple way I can use this in my existing code right now?

I tried with `source = "github.com/nauxliu/terraform-aws-eks.git?ref=feature/0.12"` but it looks like TF chokes when you have a branch containing '/'.

**Comment by endorama** [comment]
@cawilliamson this code works for me:

```
module "ci-cluster" {
  # source  = "terraform-aws-modules/eks/aws"
  # version = "4.0.2"
  source = "git::https://github.com/nauxliu/terraform-aws-eks.git?ref=feature/0.12"
  
  [...]
}
```

Ensure you commented out the `version` if any :) First time I tried I just changed `source`, without removing `version` and terraform was complaining.

**Comment by mathuin** [comment]
@nauxliu Everything appears to be working!

**Comment by rbayliss** [comment]
This branch worked perfectly for met to spin up a new cluster.  I didn't exercise all of the different options, but the basic use case is definitely working.

**Comment by endorama** [comment]
I can confirm it's working here too!

@nauxliu nice work!

**Comment by krjackso** [comment]
Used this branch to upgrade to Terraform 0.12, felt some uneasiness over these lines in the plan:

module.k8s1.module.eks_cluster.local_file.config_map_aws_auth[0] must be replaced
module.k8s1.module.eks_cluster.null_resource.update_config_map_aws_auth[0] must be replaced

But everything worked out okay.

**Comment by heydonovan** [comment]
Make sure to use double-slashes if using Terragrunt:

```
module "ci-cluster" {
  # source  = "terraform-aws-modules/eks/aws"
  # version = "4.0.2"
  source = "git::https://github.com/nauxliu/terraform-aws-eks.git//?ref=feature/0.12"
  
  [...]
}
```

Was originally receiving this error:

```
Initializing modules...
Downloading git::https://github.com/nauxliu/terraform-aws-eks.git?ref=feature/0.12 for eks...

Error: Invalid version constraint

Cannot apply a version constraint to module "eks" (at main.tf:10) because it
has a non Registry URL.
```

**Comment by max-rocket-internet** [comment]
> felt some uneasiness over these lines in the plan:
> module.k8s1.module.eks_cluster.local_file.config_map_aws_auth[0] must be replaced
> module.k8s1.module.eks_cluster.null_resource.update_config_map_aws_auth[0] must be replaced

I think this is OK. I have these in version control and the contents didn't actually change.

**Comment by max-rocket-internet** [comment]
OK I've done one more test and it looks good to me. I'll merge 🚀

**Comment by kristofmartens** [comment]
Keep in mind this is a breaking change, not a big fan of forcing everybody to terraform 0.12 as not everybody can easily update to this version

**Pull request merged**